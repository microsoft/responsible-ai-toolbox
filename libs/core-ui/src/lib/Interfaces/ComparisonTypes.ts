// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

export enum ComparisonTypes {
  /**
   * (less than)
   */
  LT = "lt",
  /**
   * (less than or equal to)
   */
  LTE = "lte",
  /**
   * (greater than)
   */
  GT = "gt",
  /**
   * (greater than or equal to)
   */
  GTE = "gte",
  /**
   * (equal to)
   */
  EQ = "eq",
  /**
   * (not equal to)
   */
  NE = "ne",
  /**
   * (in the set)
   */
  IN = "in",
  /**
   * (not in the set)
   */
  NIN = "nin",
  /**
   * (int the range)
   */
  RG = "rg",
  /**
   * (not in the range)
   */
  NRG = "nrg"
}

export function flipComparisonType(
  type: ComparisonTypes
): ComparisonTypes | undefined {
  switch (type) {
    case ComparisonTypes.LT:
      return ComparisonTypes.GTE;
    case ComparisonTypes.LTE:
      return ComparisonTypes.GT;
    case ComparisonTypes.GT:
      return ComparisonTypes.LTE;
    case ComparisonTypes.GTE:
      return ComparisonTypes.LT;
    case ComparisonTypes.EQ:
      return ComparisonTypes.NE;
    case ComparisonTypes.NE:
      return ComparisonTypes.EQ;
    case ComparisonTypes.IN:
      return ComparisonTypes.NIN;
    case ComparisonTypes.NIN:
      return ComparisonTypes.IN;
    case ComparisonTypes.RG:
      return ComparisonTypes.NRG;
    case ComparisonTypes.NRG:
      return ComparisonTypes.RG;
    default:
      return undefined;
  }
}
