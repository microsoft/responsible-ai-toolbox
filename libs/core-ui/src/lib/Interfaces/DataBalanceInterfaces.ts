// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

export interface IDataBalanceMeasures {
  aggregateBalanceMeasures: IAggregateBalanceMeasures;
  distributionBalanceMeasures: IDistributionBalanceMeasures;
  featureBalanceMeasures: IFeatureBalanceMeasures;
}

export interface IAggregateBalanceMeasures {
  measures: { [measureName: string]: number };
}

export function getAggregateBalanceMeasures(measures: {
  [measureName: string]: number;
}): { [measureName: string]: number } {
  return measures;
}

export interface IDistributionBalanceMeasures {
  measures: { [featureName: string]: { [measureName: string]: number } };
}

export function getDistributionBalanceMeasures(
  measures: { [featureName: string]: { [measureName: string]: number } },
  featureName: string
): { [measureName: string]: number } {
  if (featureName in measures) {
    return measures[featureName];
  }

  return {};
}

export interface IFeatureBalanceMeasures {
  featureValues: { [featureName: string]: string[] };
  features: string[];
  measures: {
    [featureName: string]: {
      [classKey: string]: { [measureName: string]: number };
    };
  };
}

export const measureVarNames = new Map<string, string>([
  // TODO: Positive rates aren't good to visualize in a heatmap, because it is per feature value, not a combination of feature values
  ["Demographic Parity", "dp"],
  ["Jaccard Index", "ji"],
  ["Kendall Rank Correlation", "krc"],
  ["Log-Likelihood Ratio", "llr"],
  ["Normalized PMI,	p(x,y) normalization", "n_pmi_xy"],
  ["Normalized PMI,	p(y) normalization", "n_pmi_y"],
  ["Pointwise Mutual Information (PMI)", "pmi"],
  // ["Positive Rates of Class A", "prA"],
  // ["Positive Rates of Class B", "prB"],
  ["Sorensen-Dice Coefficient", "sdc"],
  ["Squared PMI", "s_pmi"],
  ["t-test", "t_test"]
]);

export const measureRanges = new Map<string, number[]>([
  ["Demographic Parity", [-1, 1]]
]);

export function getFeatureBalanceMeasures(
  measures: {
    [featureName: string]: {
      [classKey: string]: { [measureName: string]: number };
    };
  },
  featureName: string,
  classA: string,
  classB: string
): { [measureName: string]: number } {
  const classes = measures[featureName];
  if (classes) {
    let classKey = `${classA}__${classB}`;
    if (classKey in classes) {
      return classes[classKey];
    }

    classKey = `${classB}__${classA}`;
    if (classKey in classes) {
      // Measures are available for the string "classB__classA" but not for "classA__classB"
      // For certain measures, we need to inverse the measure value to get the correct values for classA and classB
      // We don't modify the underlying collection to preserve the original value, and instead return a new collection
      const measures = classes[classKey];
      const finalMeasures: { [measureName: string]: number } = {};
      for (const measure in measures) {
        switch (measure) {
          case "dp":
            finalMeasures[measure] = -1 * measures[measure];
            break;
          // case "prA":
          //   finalMeasures.prA = measures.prB;
          //   finalMeasures.prB = measures.prA;
          //   break;
          // case "prB":
          //   break;
          default:
            finalMeasures[measure] = measures[measure];
            break;
        }
      }

      return finalMeasures;
    }
  }

  return {};
}
