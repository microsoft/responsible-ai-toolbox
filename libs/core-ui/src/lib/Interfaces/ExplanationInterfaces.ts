// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

export interface IExplanationDashboardData {
  modelInformation: IModelInformation;
  dataSummary: IDatasetSummary;
  testData?: any[][];
  predictedY?: number[];
  probabilityY?: number[][];
  trueY?: number[];
  explanationMethod?: string;

  precomputedExplanations?: IPrecomputedExplanations;
}

export interface ISerializedExplanationData {
  explanationMethod?: string;
  predictedY?: any[];
  trainingData: any[][];
  isClassifier: boolean;
  trueY?: any[];
  localExplanations:
    | IMultiClassLocalFeatureImportance
    | ISingleClassLocalFeatureImportance;
  featureNames: string[];
  classNames: string[];
  probabilityY?: number[][];
  categoricalMap?: { [key: number]: string[] };
}

export interface IModelInformation {
  modelClass: "Tree" | "EBM" | "blackbox";
  method?: "classifier" | "regressor";
}

export interface IDatasetSummary {
  featureNames?: string[];
  classNames?: string[];
  categoricalMap?: { [key: number]: string[] };
}

export type IGlobalFeatureImportance =
  | IMultiClassGlobalFeatureImportance
  | ISingleClassGlobalFeatureImportance;

export interface IPrecomputedExplanations {
  localFeatureImportance?:
    | IMultiClassLocalFeatureImportance
    | ISingleClassLocalFeatureImportance;
  globalFeatureImportance?: IGlobalFeatureImportance;
  ebmGlobalExplanation?: IEBMGlobalExplanation;
  customVis?: string;
}

export interface IEBMGlobalExplanation {
  feature_list: IBoundedCoordinates[];
}

export interface IMultiClassLocalFeatureImportance {
  scores: number[][][];
  intercept?: number[];
}

export interface ISingleClassLocalFeatureImportance {
  scores: number[][];
  intercept?: number;
  featureNames?: string[];
}

export interface IMultiClassGlobalFeatureImportance {
  scores: number[][];
  intercept?: number[];
  featureNames?: string[];
}

export interface ISingleClassGlobalFeatureImportance {
  scores: number[];
  intercept?: number;
  featureNames?: string[];
}

export interface IBoundedCoordinates {
  type: string;
  names: number[] | string[];
  scores: number[] | number[][];
  scores_range?: number[];
  upper_bounds?: number[] | number[][];
  lower_bounds?: number[] | number[][];
}
