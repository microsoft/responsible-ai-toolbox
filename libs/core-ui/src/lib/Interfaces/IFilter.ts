// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

export enum FilterMethods {
  GreaterThan = "greater",
  GreaterThanEqualTo = "greater and equal",
  LessThan = "less",
  LessThanEqualTo = "less and equal",
  Equal = "equal",
  Includes = "includes",
  Excludes = "excludes",
  InTheRangeOf = "in the range of"
}

export interface IFilter {
  method: FilterMethods;
  arg: number[];
  column: string;
}

export interface IFilterContext {
  filters: IFilter[];
  onAdd: (newFilter: IFilter) => void;
  onDelete: (index: number) => void;
  onUpdate: (filter: IFilter, index: number) => void;
}

export enum Operations {
  And = "and",
  Or = "or"
}

export type ICompositeFilter =
  | IFilter
  | {
      compositeFilters: ICompositeFilter[];
      operation: Operations;

      method?: never;
    };
