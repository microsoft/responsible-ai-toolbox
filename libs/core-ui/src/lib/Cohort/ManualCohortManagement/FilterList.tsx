// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { IconButton, TooltipHost, TooltipOverflowMode } from "@fluentui/react";
import { localization } from "@responsible-ai/localization";
import { roundDecimal } from "@responsible-ai/mlchartlib";
import React from "react";

import { FilterMethods, IFilter } from "../../Interfaces/IFilter";
import { JointDataset } from "../../util/JointDataset";

export interface IFilterListProps {
  filters: IFilter[];
  jointDataset: JointDataset;
  editFilter?(index: number): void;
  removeFilter?(index: number): void;
}
export class FilterList extends React.Component<IFilterListProps> {
  private filterMethodLabels: { [key in FilterMethods]: string } = {
    [FilterMethods.Equal]: localization.Interpret.FilterOperations.equals,
    [FilterMethods.GreaterThan]:
      localization.Interpret.FilterOperations.greaterThan,
    [FilterMethods.GreaterThanEqualTo]:
      localization.Interpret.FilterOperations.greaterThanEquals,
    [FilterMethods.LessThan]: localization.Interpret.FilterOperations.lessThan,
    [FilterMethods.LessThanEqualTo]:
      localization.Interpret.FilterOperations.lessThanEquals,
    [FilterMethods.Includes]: localization.Interpret.FilterOperations.includes,
    [FilterMethods.Excludes]: localization.Interpret.FilterOperations.excludes,
    [FilterMethods.InTheRangeOf]:
      localization.Interpret.FilterOperations.inTheRangeOf
  };
  public render(): React.ReactNode {
    return this.props.filters
      .filter((item) => item)
      .map((filter, index) => {
        return (
          <div key={index}>
            {this.setFilterLabel(filter)}
            {this.props.editFilter && (
              <IconButton
                id={`editFilerBtn-${index}`}
                iconProps={{ iconName: "Edit" }}
                onClick={(): void => this.props.editFilter?.(index)}
                ariaLabel={localization.Common.editButton}
              />
            )}
            {this.props.removeFilter && (
              <IconButton
                id={`removeFilterBtn-${index}`}
                iconProps={{ iconName: "Clear" }}
                onClick={(): void => this.props.removeFilter?.(index)}
                ariaLabel={localization.Common.close}
              />
            )}
          </div>
        );
      });
  }

  private setFilterLabel(filter: IFilter): React.ReactNode {
    const selectedFilter = this.props.jointDataset.metaDict[filter.column];
    let stringArgs;
    let label = "";

    if (
      selectedFilter.isCategorical ||
      this.props.jointDataset.metaDict[filter.column]?.treatAsCategorical
    ) {
      const selectedValues: string[] = [];
      const filterArgs = filter.arg;
      filterArgs.forEach((element) => {
        const value = selectedFilter.sortedCategoricalValues?.[element];
        if (value) {
          selectedValues.push(value);
        }
      });
      stringArgs = selectedValues.toString();
      if (selectedValues.length > 3) {
        const otherValues = selectedValues.slice(0, 3).toString();
        const countOtherValues = selectedValues.length - 3;
        stringArgs = localization.formatString(
          localization.Interpret.FilterOperations.overflowFilterArgs,
          otherValues,
          countOtherValues.toString()
        );
      }
    } else {
      for (let i = 0; i < filter.arg.length; i++) {
        filter.arg[i] = roundDecimal(filter.arg[i]);
      }
      stringArgs = filter.arg.toString();
    }

    if (filter.method === FilterMethods.InTheRangeOf) {
      // example: Age [30,40]
      label = `${selectedFilter.abbridgedLabel} ${localization.formatString(
        localization.Interpret.FilterOperations.inTheRangeOf,
        stringArgs
      )}`;
    } else {
      // example: Age < 30
      label = `${selectedFilter.abbridgedLabel} ${localization.formatString(
        this.filterMethodLabels[filter.method],
        stringArgs
      )}`;
    }

    return (
      <TooltipHost
        overflowMode={TooltipOverflowMode.Self}
        content={label}
        onTooltipToggle={(): boolean => false}
      >
        {label}
      </TooltipHost>
    );
  }
}
