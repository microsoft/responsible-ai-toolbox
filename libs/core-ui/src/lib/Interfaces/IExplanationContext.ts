// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { IModelMetadata } from "@responsible-ai/mlchartlib";

import { JointDataset } from "../util/JointDataset";

export enum ModelTypes {
  Regression = "regression",
  Binary = "binary",
  Multiclass = "multiclass",
  ImageBinary = "imagebinary",
  ImageMulticlass = "imagemulticlass",
  TextBinary = "textbinary",
  TextMulticlass = "textmulticlass"
}

export interface IExplanationContext {
  modelMetadata: IExplanationModelMetadata;
  jointDataset?: JointDataset;
  explanationGenerators: IExplanationGenerators;
  localExplanation?: ILocalExplanation;
  testDataset: ITestDataset;
  globalExplanation?: IGlobalExplanation;
  isGlobalDerived: boolean;
  ebmExplanation?: IFeatureValueExplanation;
  customVis?: string;
  inputError?: string;
}

// The interface containing either the local explanations matrix,
// or information on the fetching of the local explanation.
export interface ILocalExplanation {
  values: number[][][];
  flattenedValues?: number[][];
  intercepts?: number[];
  percentComplete?: number;
}

// The Global explanation. Either derived from local, or passed in as independent prop.
// User provided shall take precidence over our computed. Features x Class
export interface IGlobalExplanation {
  perClassFeatureImportances?: number[][];
  flattenedFeatureImportances?: number[];
  intercepts?: number[];
}

export interface IMultiClassBoundedCoordinates {
  type: string;
  names: number[] | string[];
  scores: number[][];
  scoresRange?: number[];
  upperBounds?: number[][];
  lowerBounds?: number[][];
}

export interface IFeatureValueExplanation {
  featureList: Array<IMultiClassBoundedCoordinates | undefined>;
  displayParameters?: {
    interpolation?: "vh" | "other";
    yAxisLabel?: string;
    xAxisLabel?: string;
  };
}

export interface IExplanationGenerators {
  requestPredictions?: (
    request: any[],
    abortSignal: AbortSignal
  ) => Promise<any[]>;
  requestLocalFeatureExplanations?: (
    request: any[],
    abortSignal: AbortSignal,
    explanationAlgorithm?: string
  ) => Promise<any[]>;
}

export interface ITestDataset {
  dataset?: any[][];
  predictedY?: number[];
  probabilityY?: number[][];
  trueY?: number[];
}

export interface IExplanationModelMetadata extends IModelMetadata {
  modelType: ModelTypes;
  explainerType?: string;
}
