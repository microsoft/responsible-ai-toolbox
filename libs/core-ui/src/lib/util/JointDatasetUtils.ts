// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { INumericRange } from "@responsible-ai/mlchartlib";

import {
  IMultiClassLocalFeatureImportance,
  ISingleClassLocalFeatureImportance
} from "../Interfaces/ExplanationInterfaces";
import { IExplanationModelMetadata } from "../Interfaces/IExplanationContext";
import { IFeatureMetaData } from "../Interfaces/IMetaData";

import { AxisTypes } from "./IGenericChartProps";

export interface IJointDatasetArgs {
  dataset?: any[][];
  predictedY?: number[];
  predictedProbabilities?: number[][];
  trueY?: number[];
  localExplanations?:
    | IMultiClassLocalFeatureImportance
    | ISingleClassLocalFeatureImportance;
  metadata: IExplanationModelMetadata;
  featureMetaData?: IFeatureMetaData;
}

export enum ColumnCategories {
  Outcome = "outcome",
  Dataset = "dataset",
  Index = "index",
  Explanation = "explanation",
  Cohort = "cohort",
  None = "none"
}

export enum ClassificationEnum {
  TrueNegative = 0,
  FalsePositive = 1,
  FalseNegative = 2,
  TruePositive = 3
}

export enum MulticlassClassificationEnum {
  Correct = 0,
  Misclassified = 1
}

// The object that will store user-facing strings and associated metadata
// It stores the categorical labels for any numeric bins
export interface IJointMeta {
  label: string;
  abbridgedLabel: string;
  AxisType?: AxisTypes;
  isCategorical: boolean;
  // used to allow user to treat integers as categorical (but switch back as convenient...)
  treatAsCategorical?: boolean;
  sortedCategoricalValues?: string[];
  featureRange?: INumericRange;
  category: ColumnCategories;
  index?: number;
}

export interface IDatasetMeta {
  featureMetaData?: IFeatureMetaData;
}
