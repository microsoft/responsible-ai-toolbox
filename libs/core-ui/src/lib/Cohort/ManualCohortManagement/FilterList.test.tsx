// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { shallow } from "enzyme";
import React from "react";

import { IExplanationModelMetadata } from "../../Interfaces/IExplanationContext";
import { FilterMethods, IFilter } from "../../Interfaces/IFilter";
import { JointDataset } from "../../util/JointDataset";
import { ColumnCategories } from "../../util/JointDatasetUtils";

import { FilterList } from "./FilterList";

const featureIsCategorical = [
  false,
  false,
  false,
  true,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false
];
const modelMetadata = {
  classNames: ["Class 0"],
  featureIsCategorical,
  featureNames: [],
  featureNamesAbridged: [],
  featureRanges: [],
  modelType: "binary"
} as IExplanationModelMetadata;
const jointDataset = new JointDataset({
  dataset: [],
  metadata: modelMetadata
});

describe("FilterList", () => {
  it("Should render", () => {
    const filters = [
      {
        arg: [12.09],
        column: "alcohol",
        method: FilterMethods.LessThan
      },
      {
        arg: [2.5],
        column: "ash",
        method: FilterMethods.GreaterThan
      }
    ];

    jointDataset.metaDict.alcohol = {
      abbridgedLabel: "l",
      category: ColumnCategories.Dataset,
      isCategorical: true,
      label: "less than",
      treatAsCategorical: true
    };
    jointDataset.metaDict.ash = {
      abbridgedLabel: "g",
      category: ColumnCategories.Dataset,
      isCategorical: true,
      label: "greater than",
      treatAsCategorical: true
    };

    const wrapper = shallow(
      <FilterList filters={filters} jointDataset={jointDataset} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("Should filter out undefined filter", () => {
    const filters = [
      {
        arg: [12.09],
        column: "alcohol",
        method: FilterMethods.LessThan
      },
      {
        arg: [2.5],
        column: "ash",
        method: FilterMethods.GreaterThan
      },
      undefined as unknown as IFilter
    ];

    jointDataset.metaDict.alcohol = {
      abbridgedLabel: "l",
      category: ColumnCategories.Dataset,
      isCategorical: true,
      label: "less than",
      treatAsCategorical: true
    };
    jointDataset.metaDict.ash = {
      abbridgedLabel: "g",
      category: ColumnCategories.Dataset,
      isCategorical: true,
      label: "greater than",
      treatAsCategorical: true
    };

    const wrapper = shallow(
      <FilterList filters={filters} jointDataset={jointDataset} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
