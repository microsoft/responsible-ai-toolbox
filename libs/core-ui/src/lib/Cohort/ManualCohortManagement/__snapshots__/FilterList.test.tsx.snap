// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterList Should filter out undefined filter 1`] = `
Array [
  <div
    key="0"
  >
    <StyledTooltipHostBase
      content="l  less than "
      onTooltipToggle={[Function]}
      overflowMode={1}
    >
      l  less than 
    </StyledTooltipHostBase>
  </div>,
  <div
    key="1"
  >
    <StyledTooltipHostBase
      content="g  greater than "
      onTooltipToggle={[Function]}
      overflowMode={1}
    >
      g  greater than 
    </StyledTooltipHostBase>
  </div>,
]
`;

exports[`FilterList Should render 1`] = `
Array [
  <div
    key="0"
  >
    <StyledTooltipHostBase
      content="l  less than "
      onTooltipToggle={[Function]}
      overflowMode={1}
    >
      l  less than 
    </StyledTooltipHostBase>
  </div>,
  <div
    key="1"
  >
    <StyledTooltipHostBase
      content="g  greater than "
      onTooltipToggle={[Function]}
      overflowMode={1}
    >
      g  greater than 
    </StyledTooltipHostBase>
  </div>,
]
`;
